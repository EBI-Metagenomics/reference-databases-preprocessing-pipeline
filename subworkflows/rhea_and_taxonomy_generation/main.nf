include { UNIREF90_RHEA_FILTER      } from '../../modules/local/uniref90_rhea_filter/main.nf'
include { UNIREF90_NON_VIRAL_FILTER } from '../../modules/local/uniref90_non_viral_filter/main.nf'
include { REFORMAT_RHEA_CHEBI       } from '../../modules/local/reformat_rhea_chebi/main.nf'
include { EXTRACT_TAXDUMP           } from '../../modules/local/extract_taxdump/main.nf'
include { DIAMOND_MAKEDB            } from '../../modules/nf-core/diamond/makedb/main.nf'
include { CATPACK_PREPARE           } from '../../modules/nf-core/catpack/prepare/main.nf'

workflow RHEA_AND_TAXONOMY_GENERATION {
    take:
        uniref90_fasta
        uniprot_rhea_mapping
        txt_rhea_chebi_mapping
        ncbi_taxdump

    main:
        UNIREF90_RHEA_FILTER(uniref90_fasta, uniprot_rhea_mapping)

        UNIREF90_RHEA_FILTER.out.filtered_proteins
            .map { filepath -> 
                [[id: "uniref90_rhea_${params.uniref90_version}_${params.uniprotKB_access_date}"], filepath]
            }
            .set { uniref90_with_rhea_ch }
        
        DIAMOND_MAKEDB(uniref90_with_rhea_ch, [], [], [])

        Channel.fromPath(uniref90_fasta)
            .splitFasta(by: params.uniref90_batch_size, file: 'uniref90')
            .set { uniref90_batches_ch }

        UNIREF90_NON_VIRAL_FILTER(uniref90_batches_ch)

        UNIREF90_NON_VIRAL_FILTER.out.filtered_proteins
            .collectFile(name: 'uniref90_non_viral.fasta')
            .map { filepath ->
                [[id: "uniref90_taxonomy_${params.uniref90_version}"], filepath]
            }
            .set { uniref90_non_viral_ch }

        UNIREF90_NON_VIRAL_FILTER.out.protid2taxid
            .collectFile(name: 'uniref90_non_viral.protid2taxid')
            .set { uniref90_non_viral_mapping_ch }

        EXTRACT_TAXDUMP(ncbi_taxdump)

        CATPACK_PREPARE(
            uniref90_non_viral_ch,
            EXTRACT_TAXDUMP.out.tax_names,
            EXTRACT_TAXDUMP.out.tax_nodes,
            uniref90_non_viral_mapping_ch
        )

        REFORMAT_RHEA_CHEBI(txt_rhea_chebi_mapping)

    emit:
        rhea_db            = DIAMOND_MAKEDB.out.db
        cat_diamond_db     = CATPACK_PREPARE.out.db
        cat_taxonomy_db    = CATPACK_PREPARE.out.taxonomy
        rhea_chebi_mapping = REFORMAT_RHEA_CHEBI.out.tsv_rhea_chebi_mapping
}
