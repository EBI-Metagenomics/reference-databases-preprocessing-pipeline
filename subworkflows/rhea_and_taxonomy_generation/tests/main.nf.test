nextflow_workflow {

    name "Test workflow RHEA_AND_TAXONOMY_GENERATION"
    script "../main.nf"
    workflow "RHEA_AND_TAXONOMY_GENERATION"
    config './nextflow.config'

    test("Run RHEA_AND_TAXONOMY_GENERATION module on test dataset") {

        when {
            workflow {
                """
                input[0] = "${moduleDir}/tests/data/uniref90.test.fasta"
                input[1] = "${moduleDir}/tests/data/uniprot_rhea_mapping.test.tsv"
                input[2] = "${moduleDir}/tests/data/rhea-reactions.txt.gz"
                input[3] = "${moduleDir}/tests/data/taxdump.tar.gz"
                """
            }
            params {
                    uniref90_version = "2024_05"
                    rheadb_version = "135"
                    uniprotKB_access_date = "2024-07-31"
                    outdir = "${outputDir}"
            }
        }

        then {
            assertAll(
            { assert workflow.success },
            { assert snapshot(workflow.out).match() },
            )
        }
    }
}
