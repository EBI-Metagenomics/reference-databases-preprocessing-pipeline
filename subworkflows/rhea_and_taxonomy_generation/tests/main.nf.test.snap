{
    "rhea_db": {
        "content": [
            [
                [
                    {
                        "id": "uniref90_rhea_2024_05_2024-07-31"
                    },
                    "uniref90_rhea_2024_05_2024-07-31.dmnd:md5,24d32b7dfd8efa693ddb182d34465070"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-02-13T11:37:09.175032"
    },
    "Run the workflow on the test dataset": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "uniref90_rhea_2024_05_2024-07-31"
                        },
                        "uniref90_rhea_2024_05_2024-07-31.dmnd:md5,d7c104b6576ae8a796066aef4937d339"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "2024_05"
                        },
                        [
                            "2025-02-14_CAT_pack.dmnd:md5,b72362dc62c390ea34714d48430dfd59",
                            "2025-02-14_CAT_pack.fastaid2LCAtaxid:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "2025-02-14_CAT_pack.taxids_with_multiple_offspring:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "2024_05"
                        },
                        [
                            "names.dmp:md5,c471c27a4ce85ae74d2c63633c9ce1e3",
                            "nodes.dmp:md5,130f9132095562e09c732679c562f5e9"
                        ]
                    ]
                ],
                "3": [
                    "rhea_chebi_mapping_135.tsv:md5,52d55d130ac9e9329e6e32d88b335f58"
                ],
                "cat_diamond_db": [
                    [
                        {
                            "id": "2024_05"
                        },
                        [
                            "2025-02-14_CAT_pack.dmnd:md5,b72362dc62c390ea34714d48430dfd59",
                            "2025-02-14_CAT_pack.fastaid2LCAtaxid:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "2025-02-14_CAT_pack.taxids_with_multiple_offspring:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "cat_taxonomy_db": [
                    [
                        {
                            "id": "2024_05"
                        },
                        [
                            "names.dmp:md5,c471c27a4ce85ae74d2c63633c9ce1e3",
                            "nodes.dmp:md5,130f9132095562e09c732679c562f5e9"
                        ]
                    ]
                ],
                "rhea_chebi_mapping": [
                    "rhea_chebi_mapping_135.tsv:md5,52d55d130ac9e9329e6e32d88b335f58"
                ],
                "rhea_db": [
                    [
                        {
                            "id": "uniref90_rhea_2024_05_2024-07-31"
                        },
                        "uniref90_rhea_2024_05_2024-07-31.dmnd:md5,d7c104b6576ae8a796066aef4937d339"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-02-14T15:48:28.618029"
    },
    "cat_taxonomy_db": {
        "content": [
            [
                [
                    {
                        "id": "uniref90_taxonomy_2024_05"
                    },
                    [
                        "names.dmp:md5,c471c27a4ce85ae74d2c63633c9ce1e3",
                        "nodes.dmp:md5,130f9132095562e09c732679c562f5e9"
                    ]
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-02-13T15:09:30.441085"
    },
    "cat_diamond_db": {
        "content": [
            [
                [
                    {
                        "id": "uniref90_taxonomy_2024_05"
                    },
                    [
                        "2025-02-13_CAT_pack.dmnd:md5,b72362dc62c390ea34714d48430dfd59",
                        "2025-02-13_CAT_pack.fastaid2LCAtaxid:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "2025-02-13_CAT_pack.taxids_with_multiple_offspring:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-02-13T15:09:30.428836"
    },
    "rhea_chebi_mapping": {
        "content": [
            [
                "rhea_chebi_mapping_135.tsv:md5,52d55d130ac9e9329e6e32d88b335f58"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-02-13T11:37:09.19346"
    }
}